econsa/correlation.py

```diff
@@ -0,0 +1,156 @@
+    >>> np.testing.assert_almost_equal(corr, corr_copula, decimal=1)
+    """
+    corr = np.atleast_2d(corr)
+
+    # Test that marginals are all cp.distributions.
+    for marginal in marginals:
+        if issubclass(type(marginal), cp.distributions.baseclass.Dist):
+            continue
+        else:
+            raise NotImplementedError("marginals must be chaospy distributions")
+
+    if check_corr:
+        if not np.allclose(corr, corr.T):
+            raise ValueError("corr is not a symmetric matrix")
+        elif not np.all((corr >= -1) & (corr <= 1)):
+            raise ValueError("corr must be between 0 and 1")
+        elif not np.all(np.diagonal(corr) == 1):
+            raise ValueError("the diagonal of corr must all be 1")
+        elif np.all(np.linalg.eigvals(corr) > 0) == 0:
+            raise ValueError("corr is not positive-definite")
+
+    dim = len(corr)
+    indices = np.tril_indices(dim, -1)
+    gc_corr = np.identity(dim)
+
+    for i, j in list(zip(*indices)):
+        subset = [marginals[i], marginals[j]]
+        distributions, rho = cp.J(*subset), corr[i, j]
+        gc_corr[i, j] = _gc_correlation_pairwise(distributions, rho)
+
+    # Align upper triangular with lower triangular.
+    gc_corr = gc_corr + gc_corr.T - np.diag(np.diag(gc_corr))
+
+    return gc_corr
+
+
+def _gc_correlation_pairwise(distributions, rho, seed=123, num_draws=100000):
+
+    assert len(distributions) == 2
+
+    # Test whether [L1986]_ is applicable
+    # Extract types of distributions
+    dist_type = list()
+    for dist in distributions:
+        dist_type.append(str(dist).split(sep="(", maxsplit=1)[0].lower())
+
+    try:
+        dist_norm = dist_type.index("normal")
+        dist_type.pop(dist_norm)
+        dist_other = dist_type[0]
+
+        # TODO: the code here only includes table 4,
+        # we want to also also incorporate table 5.
+        if "uniform" == dist_other:
+            f = 1.023
+        elif "exponential" == dist_other:
+            f = 1.107
+        elif "rayleigh" == dist_other:
+            f = 1.014
+        elif "logweibull" == dist_other:
+            # Type-I extreme value, Gumbel
+            f = 1.031
+        else:
+            raise ValueError("This combination is not implemented.")
+
+        result = rho * f
+    except ValueError:
+        arg_1 = np.prod(cp.E(distributions))
+        arg_2 = np.sqrt(np.prod(cp.Var(distributions)))
+        arg = rho * arg_2 + arg_1
+
+        kwargs = dict()
+        kwargs["args"] = (arg, distributions, seed, num_draws)
+        kwargs["bounds"] = (-0.99, 0.99)
+        kwargs["method"] = "bounded"
+
+        out = optimize.minimize_scalar(_criterion, **kwargs)
+        assert out["success"]
+        result = out["x"]
+
+    return result
+
+
+def _criterion(rho_c, arg, distributions, seed, num_draws):
+
+    cov = np.identity(2)
+    cov[1, 0] = cov[0, 1] = rho_c
+
+    np.random.seed(seed)
+
+    # TODO: Here we need to use proper quadrature rules
```

Comment:
TODO, the comment above,  this is already an issue and so we can just remove it.